### PR TITLE
Fix bug where kubectl completion zsh returns nonzero exit code

### LINKED ISSUE #125: Kubectl completion zsh error
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): Kubectl completion zsh

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.2", GitCommit:"bdaeafa71f6c7c04636251031f93464384d54963", GitTreeState:"clean", BuildDate:"2017-10-24T19:48:57Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.8-gke.0", GitCommit:"a7061d4b09b53ab4099e3b5ca3e80fb172e1b018", GitTreeState:"clean", BuildDate:"2017-10-10T18:48:45Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**: Cloud
- **OS** (e.g. from /etc/os-release): Mac OS High Sierra 10.13.2 Beta
- **Kernel** (e.g. `uname -a`):
- **Install tools**: 
Iterm2
shell zsh 5.2
- **Others**:

**What happened**:
source <(Kubectl completion zsh) -> returns a SIG 127 when .zshrc completes
The autocomplete seems to work but I get an error every time I start an Iterm2 zsh terminal

**What you expected to happen**:
No error

**How to reproduce it** (as minimally and precisely as possible):
source <(Kubectl completion zsh)

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by brianpursley**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
When you source zsh completion from kubectl, the exit code is nonzero which makes it seem like it failed.  This PR removes the attempted call of `_complete` which was silently failing (stderr was being redirected to /dev/null).

Example of the problem before this PR:
```
$ source <(kubectl completion zsh)
$ echo $?
127
```

Example after this PR:
```
$ source <(kubectl completion zsh)
$ echo $?
0
```

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubectl/issues/125

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Fixed bug where a nonzero exit code was returned when initializing zsh completion even though zsh completion was successfully initialized
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by antoineco** [comment]
@brianpursley could you please clarify the reason why `_complete` was failing? What was the error?

**Comment by brianpursley** [comment]
@antoineco Sure, sorry about that.

The error is suppressed, but removing the `2> /dev/null` from the generated script reveals this error:
```
./completion.sh:12464: command not found: _bash_comp
```

The additional context is in a [long-lived kubectl issue](https://github.com/kubernetes/kubectl/issues/125) where the workaround was to [use sed to comment out the alias on the fly](https://github.com/kubernetes/kubectl/issues/125#issuecomment-578523586) or [remove the _complete call altogether](https://github.com/kubernetes/kubectl/issues/125#issuecomment-448427166).

Removing the alias works for some people (mac maybe?), but for me, removing the alias alone was not enough because I then got different errors:
```
(eval):1: no matches found: *:globbed-files
_tags:comptags:36: can only be called from completion function
_tags:comptry:55: can only be called from completion function
_tags:comptags:60: can only be called from completion function
_tags:comptags:67: can only be called from completion function
```

So, the other suggestion in the issue comments was [whether _complete was even needed at all](https://github.com/kubernetes/kubectl/issues/125#issuecomment-448427166), since the error has been suppressed and it never really did anything anyway and completion worked fine (other than returning a failing exit code).

In my case (ubuntu 19.10, zsh 5.7.1), removing the call to _complete solved the problem.  Completion still working fine and exit code is zero.

Perhaps, I could be more defensive, leaving the call to complete in place and suppressing the exit code, like this?  As I am replying to you, I think that might be a better approach.  Simply changing this:

```
_complete kubectl 2>/dev/null
```
To this:
```
_complete kubectl 2>/dev/null || true
```

So that if it fails then it won't affect the overall exit code.

File: staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go]: masking failure here does not make sense to me. I don't have the context to know what zsh requires, but if this is not needed, why not omit it entirely?

Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go]: Thanks for the feedback @liggitt 

Yeah, my first instinct was to completely remove it because it doesn't seem to be working for anyone based on the comments in the [github issue](https://github.com/kubernetes/kubectl/issues/125). Removing it worked for me, but I was concerned because I couldn't determine what it was *supposed* to be doing in case it had some purpose I was unaware of.

I found the [commit where the line was added](https://github.com/kubernetes/kubernetes/commit/70b942d6638e99484fe9eb11e878d7300dd8d9ae), but still don't get a feel for why it was needed (or why stderr was being ignored).  I investigated zsh, bash, and cobra repos and couldn't find any mention of an implementation of `_complete` or `_bash_comp`

I guess I could go either way.  

I don't see adding `|| true` as being much worse than how it already is with `2> /dev/null`.

But on the other hand, adding `|| true` perpetuates some tech debt in that someday, someone else will come along and also have the same discussion about why it is like that and whether it can or can't be removed.

Do you have any additional context that would be helpful in deciding how to close out the issue?

If I revert back to my original commit of removing line 310 and also removing the alias on line 161, would you approve that change?

Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go]: Hi @cblecker .  Can you shed some light on why `_complete` is needed?  It has [been failing and causing the exit code to be 127](https://github.com/kubernetes/kubectl/issues/125) because _bash_comp (which it is aliased to) is not found.

I know it has been a couple of years, but the line in question is below from [this Aug 12, 2017 commit](https://github.com/kubernetes/kubernetes/commit/70b942d6638e99484fe9eb11e878d7300dd8d9ae#diff-339bcec9f0f8e5768e0f98ad47c40361R308) ...
```
_complete kubectl 2>/dev/null
```

I think this line can just be removed since it is failing and completion still seems to work, but wanted to circle back with you and make sure there is not some other specific reason it is needed since it looks like you were the one who added it.

Comment (antoineco) [reply staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go]: Seconded. We should get rid of `_complete` because it is aliased to `_bash_comp`, which is *undefined* (https://github.com/kubernetes/kubectl/issues/125#issuecomment-392754413).

**Review by antoineco — CHANGES_REQUESTED** [review]
Please also remove [the alias](https://github.com/kubernetes/kubernetes/blob/1ce69f55f21d9dc458b8c6a1efce27952fd4a7ba/staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go#L161). It is not used anymore after that change.

File: staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go]: @antoineco @liggitt ptal

File: staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go]: per https://github.com/kubernetes/kubectl/issues/125#issuecomment-392754413, _bash_expand is also undefined. I don't have the context to know what should be done here, will defer to @kubernetes/sig-cli-maintainers

**Comment by liggitt** [comment]
/unassign

/assign @seans3 
can you take a look at this?

**Review by cblecker — COMMENTED** [review]
This seems fine to me. If I remember correctly, this originally had something to do with the way zsh completions were auto generated from bash, and how they worked with the compinit framework. However, I tested with these lines removed, and it works fine in zsh 5.8.

File: staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go]: confirmed and tested locally.  Removed _expand alias.

**Comment by brianpursley** [comment]
/cc @antoineco
/cc @seans3 

I'm following up on my open PRs... Still looking for approval on this one.

It fixes a long-standing issue with zsh completion.

There has been lots of discussion, but I believe the consensus is that removal of these lines is OK, and it fixes the issue.

**Comment by antoineco** [comment]
I am not an approver, but I share your opinion.

**Comment by pwittrock** [comment]
Guess I am not either :)

**Comment by antoineco** [comment]
@cblecker can you help here maybe?

**Comment by cblecker** [comment]
I pinged Sean in slack. Hopefully he'll be able to look soon

**Comment by seans3** [comment]
/lgtm

Thanks @brianpursley

**Pull request merged**
